### ISSUE TITLE
A backstop to prevent buffering excessive data?

### ISSUE THREAD

**Issue by domenic** [domenic/streaming-mediastreams#6]
One of the beautiful features of streams is that they buffer all their chunks into an internal queue for later reading. This way, authors can react to and use the data at a pace appropriate for their application, as opposed to an event-based system where either the author has to build a buffering system manually, or they must react to the data at the pace determined by the source.

This is generally a good thing for media recording: nobody wants to lose recorded data, as could easily happen in an event-based system. But it does mean that there's a potential to buffer an unlimited amount of data. Should we consider a backstop mechanism to prevent this? I see two potential designs, both as options to the `MediaStreamRecorder` constructor:
- `maxBytes: 10 * 1024 * 1024` and/or `maxTime: 5 * 60`. When that much total data has been recorded over the entire lifetime of the stream, the recording automatically stops (with the stream then becoming closed). This might fit well with certain use cases anyway.
- `maxBufferedBytes: 10 * 1024 * 1024`. This indicates that if the internal queue contains more than the given number of bytes, then recording pauses, and will automatically resume some time after the author starts draining the queue to free up space.

Or we can say that this is not a real problem, and that recording a large video file is tantamount to using FileReader on a large Blob or calling `response.arrayBuffer()` on the fetch response for that video file or similar, and you deserve what you get.

**Comment by martinthomson** [comment]
Real-time media is going to make this necessary.  In that case, you can't just apply back pressure.  You have to start throwing away data at some point.  Values - or at least maximum values - for these will need to be set by the UA.

**Comment by domenic** [comment]
Yeah, backpressure does not work for this use case. But I am curious why you feel this is more necessary than, say, disallowing people from doing XHRs for 1 GiB files, or similar. We don't set such limits on other platform APIs that are allowed to read arbitrary amounts of data into memory.

**Comment by martinthomson** [comment]
Oh, there are limits, it's just that you aren't going to see them without actually trying.  And likely, those limits are enforced by crashing :)

**Comment by domenic** [comment]
Yeah. I guess then the question is whether the spec would mention such a limit for MediaStream recording, and/or allow it to be controlled, versus just leaving it implementation-defined like the existing limits are.
